tests/rust/safe_abstraction/mutex_u32.rs

```diff
@@ -0,0 +1,444 @@
+                close full_borrow_wrapper(kmlong, u32_full_borrow_content(t, &(*lock).data))();
+                close sep(MutexGuardU32_fbc_rest(km, kmlong, t, self, lock), full_borrow_wrapper(kmlong, u32_full_borrow_content(t, &(*lock).data)))();
+                close_full_borrow_strong(kstrong, MutexGuardU32_full_borrow_content0(km, t, self),
+                    sep(MutexGuardU32_fbc_rest(km, kmlong, t, self, lock), full_borrow_wrapper(kmlong, u32_full_borrow_content(t, &(*lock).data))));
+                full_borrow_split(kstrong, MutexGuardU32_fbc_rest(km, kmlong, t, self, lock),
+                    full_borrow_wrapper(kmlong, u32_full_borrow_content(t, &(*lock).data)));
+                full_borrow_unnest(kstrong, kmlong, u32_full_borrow_content(t, &(*lock).data));
+                lifetime_inclusion_glb(a, kstrong, kmlong);
+                full_borrow_mono(lifetime_intersection(kstrong, kmlong), a, u32_full_borrow_content(t, &(*lock).data));
+            }
+        @*/
+        //@ leak full_borrow(kstrong, _);
+        r
+    }
+    /* Todo: Since the single parameter of `drop` is a mutable reference, when we open the full borrow and destroy the meaning of `[[T]].OWN`
+    we will not be able to close borrow again. It is fine because the value is getting dropped and will not get used after our borrow lifetime ends.
+    However, to get our lifetime back we need to close the borrow wich after destroying the `Own` predicate is not always possible.
+    One way to handle this is to generate a different contract for `drop` implementations which does not return the lifetime token corresponding to lifetime
+    of the mutable reference. It will represent the fact that we know for this special case, i.e. `drop` this external lifetime will immediately end
+    after this function call and the original value will not be used afterward too.
+    */
+    // Todo: It should be an `impl` of `Drop` and a safe function
+    unsafe fn drop<'a>(&'a mut self)
+    /*@ req thread_token(?t) &*& [?qa]lifetime_token(?a) &*& exists(?km)
+        &*& full_borrow(a, MutexGuardU32_full_borrow_content0(km, t, self)) &*& lifetime_inclusion(a, km) == true;
+    @*/
+    //@ ens thread_token(t) /* &*& [qa]lifetime_token(a) */; // read the comment above
+    {
+        //@ open_full_borrow(qa/2, a, MutexGuardU32_full_borrow_content0(km, t, self));
+        //@ open MutexGuardU32_full_borrow_content0(km, t, self)();
+        //@ open MutexGuardU32_own(km)(t, ?lock);
+        //@ assert [_]exists_np(?kmlong);
+        //@ lifetime_token_trade(a, qa/2, km);
+        //@ assert [?qkm]lifetime_token(km);
+        //@ open_frac_borrow(km, MutexU32_frac_borrow_content(kmlong, t, lock), qkm);
+        //@ open MutexU32_frac_borrow_content(kmlong, t, lock)();
+        unsafe {
+            (*self.lock).inner.unlock();
+        }
+        //@ assert [?qp]SysMutex_share(&(*lock).inner, _);
+        //@ close [qp]MutexU32_frac_borrow_content(kmlong, t, lock)();
```

Comment:
With the spec you wrote, there's nothing forcing you to close the frac borrow here. Which shows that the spec you wrote needs work.